### PR TITLE
remove custom metaclass

### LINKED ISSUE #36: remove custom metaclass?
Would be nice if we could get rid of the custom metaclass, as a custom metaclass puts a major burden on later class inheritance. The metaclass does not much beside collecting computed fields and their models into a container the auto resolver can operate on to create the resolver map. Beside that, it only holds static methods and graph attributes, that easily could be hosted on module level or another aggregator object.
 
Still needs some investigation:
- Proper field propagation possible, even with abstract models?
- Do we need `contribute_to_class`?
- Other limitations? Fixtures? Migrations?

### PR DESCRIPTION
**Pull request by jerch**
Trying to get rid of the metaclass. Shall close #36.

TODO:
- [x] remove metaclass
- [x] reshape resolver code, prolly into own module
- [x] passes all tests
- [x] abstract models keep working
- [x] check fixture and migration behavior
- [x] update docs
- [x] coverage for resolver.py

If this lands, we can prolly enter beta versioning. Imho it is the last major pending issue with the potential to turn things upside-down.

~By removing the metaclass it turned out, that the only custom work `ComputedFieldsModel` does, is the cf calculations in the `save` overload. But with the recent changes that can be decoupled easily, thus we also should investigate, whether we can drop the custom model as well. This would lower the API to just placing `@computed` decorators on model methods in question.~ Removing the custom model class `ComputedFieldsModel` turned out as a bad idea, as late resyncing of computed values cannot be done easily anymore. Thus `ComputedFieldsModel` stays in place.

### CODE REVIEW

**Comment by jerch** [comment]
Edit: Stripped the logic from `ComputedFieldsModel` and reworked the custom save logic as a save decorator on models with `@computed` instead. `ComputedFieldsModel` is now deprecated.

All it needs to declare a computed field now is `@computed` on some model method:
```python
from computedfields.models import computed

class MyModel(models.Model):
    @computed(SomeField(...), depends=[...])
    def comp(self):
        return ...
```

File: computedfields/resolver.py [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+        computed_models = {}
+        for model, computedfields in ACTIVE_COLLECTOR.models_with_computedfields:
+            if not computedfields:
+                continue
+            computed_models[model] = {}
+            _computed_fields = {}       # FIXME: remove from codebase
+            for field in computedfields:
+                computed_models[model][field.name] = field._computed['depends']
+                _computed_fields[field.name] = field
+            model.save = decorate_save(model.save)
```
Comment (mobiware) [inline computedfields/resolver.py:286]: Why did you choose this solution vs. registering a `post_save` signal on all models that have computed fields ?

File: computedfields/resolver.py [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+        returning (model, list_of_computedfields).
+        """
+        if not self.sealed:
+            raise ResolverException('collector must be sealed before accessing model field associations')
+        for model in self.models:
+            fields = set()
+            for field in model._meta.fields:
+                if field in self.computedfields:
+                    fields.add(field)
+            yield (model, fields)
```
[the remark this answers is not in the document]
Comment (jerch) [reply computedfields/resolver.py:61]: Well thats a result of the `Collector` separation from the resolver and the idea, that a resolver should know all models with its computed fields state. I thought I need this separation but it turns out, that this is just bloatware. My early attempt had the collector functionality on the resolver itself, and I think it should be moved there again. So this field is still in major transition (did not find the proper interface yet).

File: computedfields/resolver.py (hunk 1 above)
Comment (jerch) [reply computedfields/resolver.py:286]: Thats coupled to an issue, that I never phrased into an issue report. In the old code `save` overwrites will still hold dirty cf values, which basically makes them unusable for custom handling in your own `save` method. The decorator pattern solves this.

Not implemented yet: If the decorator finds a custom save method it shall do the final cf calcs again afterwards. Not sure yet about that one, maybe it is easier to skip that (perfwise) and place a note in the docs, how to sync cf values after own field modifications in save.

Overall this is the only way I could find to make cf values usable in save and ensure that the DB still gets synced values.

Note that `post_save` is to late to do that (values already written to DB), unless the record gets saved a second time.

**Comment by jerch** [comment]
@mobiware Gonna remove most of the indirection planned for custom resolvers for now as it just bloats the code and I still have no clear idea how to shape and use that later on. It also would only be beneficial for highly customized model edge cases (like on-the-fly model creation etc.) - nothing to deal with for the standard use cases. It def. should not block the removal of the metaclass.

**Comment by jerch** [comment]
@mobiware About the custom `save` overloads and whether to deal there with already updated computed field values:

I gonna restore `ComputedFieldsModel` with a custom save method again - from data integrity perspective it is better not to use the decorator pattern on save methods in the first place, but to do the value calculations right before it lands in the DB. This way it is guaranteed to contain sync values. The decorator pattern opens the door to desync issues if someone forgets to resync values after manual changes in a custom save method.

Furthermore if someone still wants to use freshly updated computed fields in a custom save method, it is easier to document to call `update_computedfields` prehand before accessing those values in the custom save method.

**Comment by mobiware** [comment]
@jerch the `@precomputed` decorator is a nice addition.

Now that you've added back `ComputedFieldsModel`, it would be great to check during the resolving process that those models that have `@computed` fields do inherit `ComputedFieldsModel`, and raise an exception if that's not the case, as it's too easy to forget to add that parent model and then scratch our heads for a while because computed fields aren't properly refreshing

**Comment by jerch** [comment]
@mobiware Eww good finding :+1: , thats indeed not covered currently.

File: computedfields/resolver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jerch) [reply computedfields/resolver.py]: Narrowed the returned models to those containing computed fields now.

**Comment by jerch** [comment]
@mobiware Imho the basic resolver rewrite is done. Would be nice if you could have a look at it. There are still some TODOs floating around, but they not mandatory (mostly nice to have things).

Missing pieces to get this public are some fixture and migration tests. Do you know an easy way how to test both with some fake data? And last but not least the docs need a serious rwrite with all those changes.

**Comment by jerch** [comment]
Fixtures now have a dedicated test case testing basic fk and fk_reverse update by `updatedata` and `update_dependent`. Also tested some migration rules in my local repo pull, still works as before.

Not sure how to automize fixture and migration testing, this gets very meta and would need tons of test scenarios to build upfront. It would be possible with https://github.com/wemake-services/django-test-migrations though. Still cumbersome to set up, maybe I might dive deeper into it with #32 in the future.

I am basically done with this PR. Would be nice to get some feedback before merging it into master.

FYI: @mobiware, @olivierdalang

File: computedfields/graph.py [hunk 3]
```diff
@@ -478,31 +423,18 @@ def resolve_dependencies(self, computed_models):
         - plus model local non computed fields, e.g. ``['self', ['fieldA', 'fieldB']]``
 
         .. warning::
-            Dont use the old `depends` notation anymore, as it is underdetermined leading to ambiguity and
-            will be removed by a later version.
+            Dont use the old `depends` notation anymore, as it is underdetermined
+            leading to ambiguity and will be removed by a later version.
```
Comment (mobiware) [inline computedfields/graph.py:427]: It has been removed now, hasn't it ?

File: computedfields/resolver.py [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+        """
+        Decorator for custom `save` methods, that expect local computed fields
+        to contain already updated values on enter.
+
+        By default local computed field values are only calculated once by the
+        `ComputedFieldModel.save` method after your own `save` method.
+
+        By placing this decorator on your save method, the values will be updated
+        before entering your method as well. Note that this comes for the price of
+        doubled local computed field calculations (before and after your save method).
```
Comment (mobiware) [inline computedfields/resolver.py:768]: Why not skip computed field calculations in `ComputedFieldsModel.saver()` if `@precomputed` is used ?

File: computedfields/graph.py (hunk 3 above)
Comment (jerch) [reply computedfields/graph.py:427]: Woops, removed.

File: computedfields/resolver.py (hunk 4 above)
Comment (jerch) [reply computedfields/resolver.py:768]: Well I currently strive for high data integrity, and we cannot guarantee this anymore, if someone does nasty late field modifications in `save` without another calculation.
Maybe this decorator should get a flag to suppress the second calculation? Something like `@precomputed(skip_second_calculation=True)`?

Comment (mobiware) [reply computedfields/resolver.py:768]: Yes a flag would definitely work

Comment (jerch) [reply computedfields/resolver.py:768]: Implemented with latest commit as `@precomputed(skip_after=True)`.

**Comment by jerch** [comment]
> What exactly do you want to test regarding migrations ? That the Resolver doesn't kick in while performing a migration ?

Well, currently the whole migration topic is treated as a SEP (someone's else problem). Computed fields just happen to work with migrations, since they are a very small abstraction on top of normal ORM fields, thus behave exactly like concrete fields (like field renaming works without additional resync, as the migration does all the needed work). Schema migrations are ugly, as they possibly invalidate everything. On the other hand changes to method code of computed fields are not covered at all by the migration logic. Currently only chance to get back to sync state is calling `manage.py updatedata` after complicated changes, but thats like rerendering the whole picture.

So basically I am unhappy about that state of non-integration, but I know not enough about migration internals beside the typical management commands. Ideally computed fields integrate more nicely with migrations (like automatically placing fine grained data migrations with some `RunPython` hooks etc).

**Comment by mobiware** [comment]
@jerch ok I understand.

It is true that, currently, "bootstraping" newly introduced computed fields on existing models requires manual operations in order to compute their initial values. However, this would probably be best addressed in another PR.

A perhaps "naive" approach to this problem would be to automatically pass a callable `default` to fields created by `@computed` decorators, and that callable would compute the field's value, but it is probably not enough since we wouldn't have any control over the order in which those `default` callables are executed, so we would probably end up with incorrect values if a new computed field depends on another new computed field on another model and its `default` callable is executed before that of its dependency.

**Comment by jerch** [comment]
@mobiware Yeah, it is def. nothing for this PR, it is more like a "first world problem" to get migrations covered more nicely.

About your idea with the default callable:
Thats one specific case, where migrations currently fall short, they need the default on the field, just to get the migration passing. But in fact they could already pull the value by running the method. But again, I know not enough of migration internals, whether during migrations access to proper model abstraction is preserved at all and such, or if a method assuming normal model means would go bonkers.

**Comment by mobiware** [comment]
@jerch well the problem is that during migrations what we're manipulating are not our actual model classes, so we can't count on `save()` calling through to `ComputedFieldsModel` own `save` method

**Comment by jerch** [comment]
> ... during migrations what we're manipulating are not our actual model classes ...

Thats what I suspected without looking into the migration handler code. Guess they do really low level things on raw SQL there. So any ORM assumptions might just break. Well, prolly better suited for its own issue/PR later on. Feel free to tackle it, if you want.

If you dont have other comments regarding this PR, I'd merge it with master. I will not yet build another release from it, so we can add late things before it goes official into 0.1.0.

**Comment by mobiware** [comment]
@jerch LGTM for merging !
